cmake/Subdirectories.cmake

```diff
@@ -155,10 +155,17 @@ function(query_gem_paths_from_external_subdirs output_gem_dirs gem_names registe
     if (gem_names)
         foreach(gem_name IN LISTS gem_names)
             unset(gem_path)
+            # Use the ERROR_VARIABLE to catch the common case when it's a simple string and not a json type.
+            string(JSON json_type ERROR_VARIABLE json_error TYPE ${gem_name})
+            set(gem_optional FALSE)
+            if(${json_type} STREQUAL "OBJECT")
+                string(JSON gem_optional GET ${gem_name} "optional")
+                string(JSON gem_name GET ${gem_name} "name")    # overwrite the gem_name variable
+            endif()
```

Comment:
We are going to have to move this extraction of the gem name entry that is an object to [get_all_external_subdirectories_for_o3de_object](https://github.com/o3de/o3de/blob/c7d31e7562c5156e11b96782a9eab931f8fcb5d7/cmake/Subdirectories.cmake#L260-L262) function below, because the `gem_names` array is iterated in that function and appended to a CMAKE property name.

What we could do there is after the `o3de_read_json_array` call to read the each gem_name field from the project.json, we can then create a "mapping" in CMake using a property that contains the gem_name in it to map to the optional value,

i.e
```cmake
o3de_read_json_array(gem_names ${object_path}/${object_json_filename} "gem_names")
foreach(gem_name IN LISTS gem_names)
    # Use the ERROR_VARIABLE to catch the common case when it's a simple string and not a json type.
    string(JSON json_type ERROR_VARIABLE json_error TYPE ${gem_name})
    set(gem_optional FALSE)
    if(${json_type} STREQUAL "OBJECT")
        string(JSON gem_optional GET ${gem_name} "optional")
        string(JSON gem_name GET ${gem_name} "name")    # overwrite the gem_name variable
    endif()

    # Set a global cmake property that stores a mapping of gem_name to optional value
    set_property(GLOBAL PROPERTY "${gem_name}_OPTIONAL" ${gem_optional})
endforeach()
```

Then in this `query_gem_paths_from_external_subdirs` function, we can just query the optional state using get_property
```cmake
    get_property(gem_optional GLOBAL PROPERTY ${gem_name}_OPTIONAL)
```

Reply:
Addressed with latest commit.